Overhaul of Parse Server Guide (#809)
ParsePlatform/Docs: _includes/parse-server/experimental.md

# Experimental

> Experimental features may not be 'battle tested' and could have issues. They are likely to become full features at some point, but could be removed from the system if they are problematic.

Reviewer: Don't know what "battle tested" means. We should be clear that experimental features are not recommended / fit for use in production systems. There is also freedom for interpretation what "problematic" means, I think the whole paragraph merits rewording.

How about:

> The following experimental features are still under active development and subject to an increased probability of bugs, breaking changes, non-backward compatible changes or removal without phased deprecation in any future version. It is not recommended to use these features in production or other critical environments.

> However, we strongly suggest and appreciate if you try out these features in development environments and provide feedback, so that they can mature faster and become production fit.

I also suggest to remove "use at your own risk" below, unless we need that for legal reasons.

Reviewer 2: Much better that my earlier suggestions! I have a couple little suggestions on top of your suggestion...

Remove `experimental` -> using the word in the definition
Remove `without phased deprecation` -> I don't think we've ever done phased deprecation so could give people the wrong idea.

> The following features are still under active development and subject to an increased probability of bugs, breaking changes, non-backward compatible changes or removal in any future version. It is not recommended to use these features in production or other critical environments.
>
> However, we strongly suggest and appreciate if you try out these features in development environments and provide feedback, so that they can mature faster and become production fit.

Reviewer: Sadly, you are right, although I am injecting phased deprecation into conversations wherever I can recently.

These features may not be approprate for production, so use at your own risk.

## Direct Access

> `directAccess` replaces the HTTP Interface when using the JS SDK in the current node runtime. This may improve performance, along with `enableSingleSchemaCache` set to `true`.

Reviewer: Not sure this description is understandable and we should mention its implications -- "replaces" with what?

How about:

> A Parse Server request in Cloud Code (such as a query) is usually sent to Parse Server via the HTTP interface, just as if the request came from outside Cloud Code and outside the current Node.js instance. This is true for all requests made with the Parse JavaScript SDK within the Node.js runtime in which Parse Server is running.
> 
> Setting `directAccess: true` instead processes requests to Parse Server directly within the Parse instance without using the HTTP interface. This also has the side effect that requests within Cloud Code cannot be distributed via a load balancer when Parse Server is running in a multi-instance environment.

* `idempotencyOptions.paths`(`Array<String>`): An array of path patterns that have to match the request path for request deduplication to be enabled. 
* The mount path must not be included, for example to match the request path `/parse/functions/myFunction` specify the path pattern `functions/myFunction`. A trailing slash of the request path is ignored, for example the path pattern `functions/myFunction` matches both `/parse/functions/myFunction` and `/parse/functions/myFunction/`.

  Examples:

  * `.*`: all paths, includes the examples below
  * `functions/.*`: all functions
  * `jobs/.*`: all jobs
  * `classes/.*`: all classes
  * `functions/.*`: all functions
  * `users`: user creation / update
  * `installations`: installation creation / update

* `idempotencyOptions.ttl`: The duration in seconds after which a request record is discarded from the database. Duplicate requests due to network issues can be expected to arrive within milliseconds up to several seconds. This value must be greater than `0`.

### Notes

- This feature is currently only available for MongoDB and not for Postgres.

## Localization

### Pages

> **Caution, this is an experimental feature that may not be appropriate for production.**

Reviewer 2: This is in the experimental section anyway...
```suggestion
```

- Localization for feature pages (e.g. password reset, email verification) is disabled if `pages.customUrls` are set, even if the custom URLs point to the pages within the pages path.
- Only `.html` files are considered for localization when localizing custom pages.

Pages can be localized in two ways:

#### Localization with Directory Structure

Pages are localized by using the corresponding file in the directory structure where the files are placed in subdirectories named after the locale or language. The file in the base directory is the default file.

**Example Directory Structure:**
```js
root/
├── public/                  // pages base path
│   ├── example.html         // default file
│   └── de/                  // de language folder
│   │   └── example.html     // de localized file
│   └── de-AT/               // de-AT locale folder
│   │   └── example.html     // de-AT localized file
```

Files are matched with the locale in the following order:
1. Locale match, e.g. locale `de-AT` matches file in folder `de-AT`.
1. Language match, e.g. locale `de-CH` matches file in folder `de`.

> 1. Default; file in base folder is returned.

Reviewer 2: ```suggestion
1. Locale match, e.g. locale `de-AT` matches file in folder `de-AT`.
2. Language match, e.g. locale `de-CH` matches file in folder `de`.
3. Default; file in base folder is returned.
```

├── private/                 // folder outside of public scope
│   └── translations.json    // JSON resource file
```

The JSON resource file loosely follows the [i18next](https://github.com/i18next/i18next) syntax, which is a syntax that is often supported by translation platforms, making it easy to manage translations, exporting them for use in Parse Server, and even to automate this workflow.

**Example JSON Content:**
```json
{
  "en": {               // resource for language `en` (English)
    "translation": {
      "greeting": "Hello!"
    }
  },
  "de": {               // resource for language `de` (German)
    "translation": {
      "greeting": "Hallo!"
    }
  }
  "de-AT": {            // resource for locale `de-AT` (Austrian German)
    "translation": {
      "greeting": "Servus!"
    }
  }
}
```

**Configuration Example:**
```js
const api = new ParseServer({
  ...otherOptions,

  pages: {
    enableRouter: true, // Enables the experimental feature; required for localization
    enableLocalization: true,
    localizationJsonPath: './private/localization.json',
    localizationFallbackLocale: 'en'
  }
}
```

Pros:

> - There is only one HTML file to maintain that contains the placeholders that are filled with the translations according to the locale.

Reviewer 2: ```suggestion
- There is only one HTML file to maintain containing the placeholders which are filled with the translations according to the locale.
```

Cons:
- Files cannot be easily previewed by viewing the file in a browser because the content contains only placeholders and even HTML or CSS changes may be dynamically applied, e.g. when a localization requires a Right-To-Left layout direction.
- Style and other fundamental layout changes may be more difficult to apply.

#### Dynamic placeholders

In addition to feature related default parameters such as `appId` and the translations provided via JSON resource, it is possible to define custom dynamic placeholders as part of the router configuration. This works independently of localization and, also if `enableLocalization` is disabled.

**Configuration Example:**
```js
const api = new ParseServer({
  ...otherOptions,

  pages: {
    enableRouter: true, // Enables the experimental feature; required for localization
    placeholders: {
      exampleKey: 'exampleValue'
    }
  }
}
```

> The placeholders can also be provided as function or as async function, with the `locale` and other feature related parameters passed through, to allow for dynamic placeholder values:

Reviewer 2: ```suggestion
The placeholders can also be provided as a function or as an async function, with the `locale` and other feature related parameters passed through, to allow for dynamic placeholder values:
```

```js
const api = new ParseServer({
  ...otherOptions,

  pages: {
    enableRouter: true, // Enables the experimental feature; required for localization
    placeholders: async (params) => {
      const value = await doSomething(params.locale);
      return {
        exampleKey: value
      };
    }
  }
}
```

#### Reserved Keys

> The following parameter and placeholder keys are reserved because they are used related to features such as password reset or email verification. They should not be used as translation keys in the JSON resource or as manually defined placeholder keys in the configuration: `appId`, `appName`, `email`, `error`, `locale`, `publicServerUrl`, `token`, `username`.

Reviewer 2: ```suggestion
The following parameter and placeholder keys are reserved because they are used in relation to features such as password reset or email verification. They should not be used as translation keys in the JSON resource or as manually defined placeholder keys in the configuration: `appId`, `appName`, `email`, `error`, `locale`, `publicServerUrl`, `token`, `username`.
```

> ### Parameters

Reviewer: Not in favor of this format, IMO harder to read than the table. Can we find something better visually structured? Edit: I am referring to the "Parameters" paragraph.

Reviewer 2: Potentially naive question - is this something that is or can be documented in the api reference?

# Experimental

Experimental Features are items that we are experimenting with that will eventually become full features, or items that will be removed from the system if they are proven to not work well.

These features may not be approprate for production, so use at your own risk.

## Direct Access

> * `directAccess`: Replaces HTTP Interface when using JS SDK in current node runtime. This may improve performance, along with `enableSingleSchemaCache` set to `true`.

Reviewer 2: ```suggestion
`directAccess` replaces the HTTP Interface when using the JS SDK in the current node runtime. This may improve performance, along with `enableSingleSchemaCache` set to `true`.
```
We also need a configuration code snippet like below.

This feature needs to be enabled on the client side to send the header and on the server to process the header. Refer to the specific Parse SDK docs to see whether the feature is supported yet.

Deduplication is only done for object creation and update (`POST` and `PUT` requests). Deduplication is not done for object finding and deletion (`GET` and `DELETE` requests), as these operations are already idempotent by definition.

Configutation:
```js
let api = new ParseServer({
    idempotencyOptions: {
        paths: [".*"],       // enforce for all requests
        ttl: 120             // keep request IDs for 120s
    }
}
```
Parameters:

* `idempotencyOptions` (`Object`): Setting this enables idempotency enforcement for the specified paths.
* `idempotencyOptions.paths`(`Array<String>`): An array of path patterns that have to match the request path for request deduplication to be enabled. 

>   * The mount path must not be included, for example to match the request path `/parse/functions/myFunction` specify the path pattern `functions/myFunction`. A trailing slash of the request path is ignored, for example the path pattern `functions/myFunction` matches both `/parse/functions/myFunction` and `/parse/functions/myFunction/`.

Reviewer 2: ```suggestion
* The mount path must not be included, for example to match the request path `/parse/functions/myFunction` specify the path pattern `functions/myFunction`. A trailing slash of the request path is ignored, for example the path pattern `functions/myFunction` matches both `/parse/functions/myFunction` and `/parse/functions/myFunction/`.
```

# Experimental

> Experimental Features are items that we are experimenting with that will eventually become full features, or items that will be removed from the system if they are proven to not work well.

Reviewer 2: ```suggestion
Experimental features may not be 'battle tested' and could have issues. They are likely to become full features at some point, but could be removed from the system if they are problematic.
```

| `idempotencyOptions`       | yes      | `Object`        | `undefined`   |                                                                                                                                                                                                                                                                             | PARSE_SERVER_EXPERIMENTAL_IDEMPOTENCY_OPTIONS | Setting this enables idempotency enforcement for the specified paths.                                                                                                                                                                                                                                                                                                                                                                                 |

> | `idempotencyOptions.paths` | yes      | `Array<String>` | `[]`          | `.*` (all paths, includes the examples below), <br>`functions/.*` (all functions), <br>`jobs/.*` (all jobs), <br>`classes/.*` (all classes), <br>`functions/.*` (all functions), <br>`users` (user creation / update), <br>`installations` (installation creation / update) | PARSE_SERVER_EXPERIMENTAL_IDEMPOTENCY_PATHS   | An array of path patterns that have to match the request path for request deduplication to be enabled. The mount path must not be included, for example to match the request path `/parse/functions/myFunction` specifiy the path pattern `functions/myFunction`. A trailing slash of the request path is ignored, for example the path pattern `functions/myFunction` matches both `/parse/functions/myFunction` and `/parse/functions/myFunction/`. |

Reviewer 2: So we have lost some of the extra info provided in this table.

@mtrezza you're better placed to comment on the importance of this

Reviewer: There have been no changes to the feature, the table's content should remain.

Reviewer 2: Maybe hold off addressing this until https://github.com/parse-community/docs/pull/809#discussion_r575697119 has been resolved
